src/main/kotlin/org/rust/lang/core/macros/builtin/BuiltinMacroExpander.kt

```diff
@@ -22,10 +28,86 @@ class BuiltinMacroExpander(val project: Project) : MacroExpander<RsBuiltinMacroD
         def: RsBuiltinMacroData,
         call: RsMacroCallData
     ): RsResult<Pair<CharSequence, RangeMap>, BuiltinMacroExpansionError> {
+        val macroBody = call.macroBody
+        if (def.name in BUILTIN_FORMAT_MACROS && macroBody is MacroCallBody.FunctionLike) {
+            val formatMacro = parseMacroBodyAsFormat(project, macroBody) ?: return Err(BuiltinMacroExpansionError)
+            val result = handleFormatMacro(def.name, macroBody.text, formatMacro) ?: return Err(BuiltinMacroExpansionError)
+            return Ok(result.text to result.ranges)
+        }
         return Err(BuiltinMacroExpansionError)
     }
 
     companion object {
-        const val EXPANDER_VERSION = 0
+        private val BUILTIN_FORMAT_MACROS: Set<String> = setOf("format_args", "format_args_nl")
+
+        const val EXPANDER_VERSION = 1
     }
 }
+
+private fun parseMacroBodyAsFormat(project: Project, macroBody: MacroCallBody.FunctionLike): RsFormatMacroArgument? {
+    val text = "(${macroBody.text})"
+
+    val (builder, _) = project
+        .createAdaptedRustPsiBuilder(text)
+        .lowerDocCommentsToAdaptedPsiBuilder(project)
+
+    val result = RustParser.FormatMacroArgument(builder, 0)
+    if (!result) return null
+
+    return builder.treeBuilt.psi as? RsFormatMacroArgument
+}
+
+private fun handleFormatMacro(macroName: String, macroText: String, format: RsFormatMacroArgument): MappedText? {
+    val formatStr = format.formatMacroArgList
+        .getOrNull(0)
+        ?.expr as? RsLitExpr
+        ?: return null
+
+    val parseCtx = parseParameters(formatStr) ?: return null
+    val errors = checkSyntaxErrors(parseCtx)
+    if (errors.isNotEmpty()) return null
+
+    val namedArguments = mutableSetOf<String>()
+    for (argument in format.formatMacroArgList) {
+        val identifier = argument.identifier
+        if (identifier != null) {
+            namedArguments.add(identifier.text)
+        }
+    }
+
+    val macroBuilder = MutableMappedText(macroText.length * 2)
+    macroBuilder.appendUnmapped("$macroName!(")
+    macroBuilder.appendMapped(macroText, 0)
+
+    fun addImplicitArgument(name: String, offset: Int) {
+        // Make sure that we have ', ' at the end before we add the new argument
+        if (!macroBuilder.text.trimEnd().endsWith(",")) {
```

Comment:
I don't like `trimEnd()` here because it copies the string (that can be large). Also it can end with a block comment after a comma. Instead we could initially figure out the comma existence by `RsFormatMacroArgument`, save it to a mutable bool variable and then update the variable value during adding new arguments

Reply:
Good idea, I changed it like that, please check it.